Source/Core/DiscIO/NFSBlob.cpp

```diff
@@ -0,0 +1,267 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "DiscIO/NFSBlob.h"
+
+#include <algorithm>
+#include <array>
+#include <cstring>
+#include <memory>
+#include <string>
+#include <string_view>
+#include <utility>
+#include <vector>
+
+#include <fmt/format.h>
+
+#include "Common/Align.h"
+#include "Common/CommonTypes.h"
+#include "Common/Crypto/AES.h"
+#include "Common/IOFile.h"
+#include "Common/Logging/Log.h"
+#include "Common/StringUtil.h"
+#include "Common/Swap.h"
+
+namespace DiscIO
+{
+NFSFileReader::NFSFileReader(NFSHeader header, std::vector<File::IOFile> files,
+                             std::array<u8, 16> key, u64 raw_size)
+    : m_header(header), m_files(std::move(files)),
+      m_aes_context(Common::AES::CreateContextDecrypt(key.data())), m_raw_size(raw_size),
+      m_data_size(CalculateExpectedDataSize(header))
+{
+}
+
+u64 NFSFileReader::CalculateExpectedRawSize(const NFSHeader& header)
+{
+  u64 total_blocks = 0;
+  for (u32 i = 0; i < Common::swap32(header.lba_range_count); ++i)
+    total_blocks += Common::swap32(header.lba_ranges[i].num_blocks);
```

Comment:
This and all the other loops over lba_ranges should have a sanity check so that they don't read out of bounds of the array. Perhaps do that when reading the header and fill your own data structure instead of just copying the file data into memory and working on that (pre-swap the big endian stuff and so on).

Reply:
Yes, that sounds like a good idea.